### PR TITLE
Mpls Path Visualization

### PR DESCRIPTION
**Pull request by vitalisator**
- Update for NOKIA MIBs
- Discovery and polling Actual Route and Computed (CSPF) Path hops
- Show route hops for active and constrain path under Mpls->Paths as drop down map
- Some marking for redundancy hops and nodes

feel free to suggest code improvements :-)

image

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by PipoCanaja** [comment]
Looks graphically good I must say. Would like to have that amount of details in my devices' mibs ... but I don't ...

File: mibs/nokia/MPLS-TE-MIB [hunk 1]
```diff
@@ -0,0 +1,1587 @@
+-- Extracted from draft-ietf-mpls-te-mib-04.txt
```
Comment (PipoCanaja) [inline mibs/nokia/MPLS-TE-MIB]: Please move this file to the ./mibs directory. This is not a nokia specific.

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-59-release-changelog-december-2019/10558/1

**Comment by willhseitz** [comment]
@vitalisator good work! What steps do you think would be needed for someone to bring this to Cisco?

**Comment by penfold1972** [comment]
While talking about getting to work on Cisco, I'd be interested in hearing what it takes to get the map on Juniper equipment.

**Comment by vitalisator** [comment]
@willhseitz , @penfold1972, I think this should be possible since the database is ready for that now. One just need MPLS information from the MIBs and some rework (copy and paste) on LibreNMS/OS/Timos.php. Unfortunately I do not own any cisco or juniper devices at this time with MPLS capability to test or implement that.

**Comment by louis-oui** [comment]
> Unfortunately I do not own any cisco or juniper devices at this time with MPLS capability to test or implement that.

Should be possible to do with virtual device on eve/GNS3

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/new-nokia-mpls-path-visualisation-issue-during-discovery-on-nokia-sar/10582/1

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/new-nokia-mpls-path-visualisation-issue-during-discovery-on-nokia-sar/10582/5

**Comment by craig-nokia** [comment]
Let me know if you need more information on this issue with the 7705 SAR.  I just altered the sql table mpls_lsp_paths to allow NULL values for mplsLspPathOperMetric.

**Comment by martijn-schmidt** [comment]
> I think this should be possible since the database is ready for that now. One just need MPLS information from the MIBs and some rework (copy and paste) on LibreNMS/OS/Timos.php. Unfortunately I do not own any cisco or juniper devices at this time with MPLS capability to test or implement that.

Hello @vitalisator - nice work on the MPLS features! I am interested in looking at the implementation for Juniper JunOS (and maybe Brocade IronWare), but since this is the first implementation I'd like to be able to see all the raw data points which the Nokia MIBs return so I have a point of reference during this work. Unfortunately my employer doesn't own any Nokia routers, and to the best of my knowledge the SNMP data collector in the master branch of LibreNMS isn't fully retrieving all relevant data right now. Would you mind repeating the test data collection process with the patch from PR #11003 applied please? Then it should be possible for me to replicate the entire Nokia device through snmpsim. Much obliged!

**Comment by penfold1972** [comment]
@vitalisator would it help if I built a Juniper lab and gave you access to it?

**Pull request merged**